pkg/mcs/port_set.go

```diff
@@ -0,0 +1,263 @@
+// the current set of ports available in the PortSet.
+func (ps *PortSet) IsCompatibleWith(rhs *PortSet) bool {
+	return !ps.IsConflictingWith(rhs)
+}
+
+// IsConflictingWith determines if the PortSet parameter has any ports
+// in conflict with the stored ports. Ports not shared by the sets are
+// ignored.
+func (ps *PortSet) IsConflictingWith(rhs *PortSet) bool {
+	return !equals(ps, rhs, validateOnlySharedPorts)
+}
+
+// Merge returns a new PortSet that holds the union of the two PortSets.
+// Returns an error if the sets are in conflict (i.e., their intersection
+// has incompatible port definitions).
+func (ps *PortSet) Merge(other *PortSet) (*PortSet, error) {
+	if ps.IsConflictingWith(other) {
+		return nil, errors.New("Conflicting")
+	}
+
+	ports := make(map[string]mcsv1a1.ServicePort, len(ps.ports))
+	for _, p := range ps.ports {
+		ports[p.Name] = p
+	}
+
+	for _, p := range other.ports {
+		if _, found := ports[p.Name]; !found {
+			ports[p.Name] = p
+		}
+	}
+
+	merged := NewPortSet()
+	for _, p := range ports {
+		merged.ports = append(merged.ports, p)
+	}
+	return merged, nil
+}
+
+// Encodes the port as an annotation. Returns the annotation name and value.
+// An error is returned if either the name or value are invalid.
+func encodePortAsAnnotation(port *mcsv1a1.ServicePort) (string, string, error) {
+	var value string
+
+	key := LIGHTHOUSE_PORT_KEY
+	if port.Name != "" {
+		key = fmt.Sprintf(LIGHTHOUSE_PORT_KEY_NAMED_FORMAT, port.Name)
+	}
+
+	if port.AppProtocol != nil && *port.AppProtocol != "" {
+		value = fmt.Sprintf(LIGHTHOUSE_PORT_VALUE_FORMAT_FULL, port.Port, port.Protocol, *port.AppProtocol)
+	} else if port.Protocol != "" {
+		value = fmt.Sprintf(LIGHTHOUSE_PORT_VALUE_FORMAT_PROTOCOL, port.Port, port.Protocol)
+	} else {
+		value = fmt.Sprintf(LIGHTHOUSE_PORT_VALUE_FORMAT_PORT, port.Port)
+	}
+
+	if err := validation.IsQualifiedName(key); len(err) > 0 {
+		return "", "", errors.New(err[0])
+	} else if err := validation.IsValidLabelValue(value); len(err) > 0 {
+		return "", "", errors.New(err[0])
+	}
+
+	return key, value, nil
+}
+
+// Decodes an annotation to determine if it matches a port specification or not.
+// Returns an error when the annotation name matches but the value can't be decoded.
+// Returns both error and port as nil when the annotation name does not match our
+// expected format.
+func decodePortFromAnnotation(name, value string) (*mcsv1a1.ServicePort, error) {
+	if strings.HasPrefix(name, LIGHTHOUSE_PORT_KEY) {
+		p := &mcsv1a1.ServicePort{}
+
+		if len(name) > len(LIGHTHOUSE_PORT_KEY) {
+			_, err := fmt.Sscanf(name, LIGHTHOUSE_PORT_KEY_NAMED_FORMAT, &p.Name)
+			if err != nil {
+				return nil, errors.New("invalid port name:" + name)
+			}
+		}
+
+		// port spec is formatted as port[-[protocol]-appPort] with [] denoting
+		// optional parts.
+		spec := strings.Split(value, "-")
+		switch len(spec) {
+		case 3:
+			p.AppProtocol = &spec[2]
```

Comment:
no validation of protocol and app protocol?

Reply:
no - they are arbitrary strings, possibly empty (in which case they wouldn't get encoded in the first place)

Reply:
it may confuse your parsing if it contains `-`

Reply:
perhaps you want the separator as a constant?

Reply:
agree about making it a constant. 
Protocol is only valid as TCP, UDP and SCTP - validated by k8s.
Since AppProtocol is arbitrary - we run the issue that it can contain any separator we think off. In practice, I haven't seen AppProtocol field used much, but I need to think about this some more (maybe encode fields with byte counts instead).
The other option is to encode each field with a unique annotation, which is simpler but also wasteful since we duplicate annotations name (`lighthouse.submariner.io/port-<field>[.name]: <value>`)
If you have better idea - I'm open to suggestions... encode json as the value?

Reply:
See later comment about reconsidering and maybe just using JSON to encode the entire Spec we need as one annotation

Reply:
Switched over to JSON encoding